### PR TITLE
 smb: New keyword smb.version v2

### PR DESCRIPTION
**Pull request by zer1t0**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [X] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [X] I have signed the Open Information Security Foundation contribution agreement at https://suricata.io/about/contribution-agreement/
- [X] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: 
https://redmine.openinfosecfoundation.org/issues/5075

Describe changes:
- New keyword smb.version added, which allows to match the packets of specif version of smb, that can be 1 or 2.
- Documentation for the keyword added (created SMB keyword documentation)

Example of rule `alert smb any any -> any any (msg: "SMBv2 message"; smb.version: 2; sid: 1;)` . More examples in documentation.

suricata-verify-pr: 740

### CODE REVIEW

File: rust/src/smb/detect.rs [hunk 1]
```diff
@@ -201,3 +204,66 @@ pub extern "C" fn rs_smb_tx_get_dce_iface(state: &mut SMBState,
...
+    let arg = match CStr::from_ptr(carg).to_str() {
+        Ok(arg) => arg,
+        _ => {
+            return std::ptr::null_mut();
+        }
+    };
+
+    match parse_version_data(arg) {
+        Ok(detect) => Box::into_raw(Box::new(detect)) as *mut _,
+        Err(_) => std::ptr::null_mut(),
```
Comment (catenacyber) [inline rust/src/smb/detect.rs:236]: can take a few less lines

File: rust/src/smb/detect.rs [hunk 2]
```diff
@@ -201,3 +204,66 @@ pub extern "C" fn rs_smb_tx_get_dce_iface(state: &mut SMBState,
...
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_smb_version_free(ptr: *mut c_void) {
+    if ptr != std::ptr::null_mut() {
```
Comment (catenacyber) [inline rust/src/smb/detect.rs:242]: @jasonish what happens if the pointer is null and we `std::mem::drop` it ?

File: rust/src/smb/detect.rs [hunk 3]
```diff
@@ -201,3 +204,66 @@ pub extern "C" fn rs_smb_tx_get_dce_iface(state: &mut SMBState,
...
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_parse_cmd_data() {
+        assert_eq!(Err(()), parse_version_data("0"));
+        assert_eq!(1u8, parse_version_data("1").unwrap());
+        assert_eq!(2u8, parse_version_data("2").unwrap());
```
Comment (catenacyber) [inline rust/src/smb/detect.rs:266]: Could we add spaces ?

File: src/Makefile.am [hunk 4]
```diff
@@ -293,6 +293,7 @@ noinst_HEADERS = \
 	detect-sip-stat-msg.h \
 	detect-sip-uri.h \
 	detect-smb-share.h \
+  detect-smb-version.h \
```
Comment (catenacyber) [inline src/Makefile.am:296]: indent is off

File: src/detect-smb-version.c [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+#include "detect-engine.h"
+#include "detect-engine-mpm.h"
+#include "detect-engine-state.h"
+#include "detect-engine-prefilter.h"
+#include "detect-engine-content-inspection.h"
+
+#include "detect-smb-version.h"
+#include "rust.h"
+
+#define PARSE_REGEX                                                                                \
```
Comment (catenacyber) [inline src/detect-smb-version.c:38]: We do not need any regex ;-)

File: src/detect-smb-version.c [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    sigmatch_table[DETECT_SMB_VERSION].name = "smb.version";
+    sigmatch_table[DETECT_SMB_VERSION].alias = "smb_version";
+
+    sigmatch_table[DETECT_SMB_VERSION].desc = "match SMB message type";
+    sigmatch_table[DETECT_SMB_VERSION].Setup = DetectSmbVersionSetup;
+    sigmatch_table[DETECT_SMB_VERSION].Match = NULL;
+    sigmatch_table[DETECT_SMB_VERSION].AppLayerTxMatch = DetectSmbVersionMatchRust;
+    sigmatch_table[DETECT_SMB_VERSION].Free = DetectSmbVersionFree;
+
+    DetectAppLayerInspectEngineRegister2(
```
Comment (catenacyber) [inline src/detect-smb-version.c:121]: Do we need these ?
I do not think so for `AppLayerTxMatch`

**Review by catenacyber — CHANGES_REQUESTED** [review]
Thanks for the changes

- CI : ✅ 
- Code : good job. Big point is the argument between C and rust for the match function...
- Commits segmentation : ok as 1
- Commit messages : could you add the ticket number in the description of the commit ?
cf https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Code_Submission_Quality_Criteria#Commits
> Information that needs to be part of a commit (if applicable): ticket it fixes. E.g. "Fixes Bug [#123](https://redmine.openinfosecfoundation.org/issues/123)."
- Git ID set : looks fine for me
- CLA : I do not have access, but looks like some commits were already merged
- Doc update : really nice  👏 
- Redmine ticket : ok
- Rustfmt : could be run as a first commit
- Tests : Suricata-verify tests look fine

File: rust/src/smb/detect.rs (hunk 2 above)
Comment (jasonish) [reply rust/src/smb/detect.rs:242]: I think its undefined, much like free'ing some invalid pointer value in C.  But I can't actually make that fail..  `Box::from_raw(null)` followed by drop doesn't seem to crash.  I think we should still null guard all these tho.  A drop impl could do some work that could then crash.

File: rust/src/smb/detect.rs (hunk 1 above)
Comment (zer1t0) [reply rust/src/smb/detect.rs:236]: What do you mean?

File: rust/src/smb/detect.rs (hunk 3 above)
Comment (zer1t0) [reply rust/src/smb/detect.rs:266]: Where?

File: src/detect-smb-version.c (hunk 6 above)
Comment (zer1t0) [reply src/detect-smb-version.c:121]: Tests fail when these are removed.

**Comment by zer1t0** [comment]
Moved to https://github.com/OISF/suricata/pull/7182

File: rust/src/smb/detect.rs (hunk 3 above)
Comment (catenacyber) [reply rust/src/smb/detect.rs:266]: like `"  2  "`

File: rust/src/smb/detect.rs (hunk 1 above)
Comment (catenacyber) [reply rust/src/smb/detect.rs:236]: do not define `let arg =`, just use it like
`Ok(arg) => match parse_version_data(arg) {`
return `std::ptr::null_mut()` only once by default

**Pull request closed without merging**